Public Course Import/Export API

bdero opened the pull request:
This is a public, shared, versioned, RESTful API for importing and exporting full course content. The code was initially ripped from the existing import/export API in the CMS contentstore djangoapp and wrapped in Django Rest Framework view classes. It's simply a new djangoapp in the *openedx* directory with 4 views:
- GET /api/v1/courses - List courses (course keys) for which the user has full author access
- GET /api/v1/courses/{course_key} - Export a full course in the form of a tarball
- POST /api/v1/courses/{course_key} - Import a full course tarball
- GET /api/v1/courses/{course_key}/import_status/{filename} - View the status of a course import (see https://github.com/edx/edx-platform/pull/6190/files#diff-4c6e0b7531e770ee217a370f8d990ac8R108)

This PR includes configuration changes. Most notably, Studio is configured to serve the OAuth2 provider alongside the LMS.

Here's the relevant thread on the code list: https://groups.google.com/forum/#!msg/edx-code/DmnHWmly25A/ZqjD1zb4o7oJ

There are 28 non-covered lines, all of which are missing coverage in the existing CMS API. They're mostly error conditions, such as handling of multipart file upload errors.

singingwolfboy commented on openedx/core/djangoapps/public_api/courses/tests/test_course_import_export.py:
```diff
@@ -0,0 +1,416 @@
+log = logging.getLogger(__name__)
+
+
+def course_url(handler, course_key, kwargs=None):
```
Don't you want this to be `**kwargs`? It's the difference between calling
```python
course_url(my_handler, my_course_key, kwargs={"foo": "bar", "num": 123})
```
and calling
```python
course_url, my_handler, my_course_key, foo=bar, num=123)
```

bdero replied:
Yep, this is much better, thanks.

singingwolfboy commented on openedx/core/djangoapps/public_api/courses/tests/test_course_import_export.py:
```diff
@@ -0,0 +1,416 @@
+        self.url = course_url('course_import_export_handler', self.course)
+        self.content_dir = path(tempfile.mkdtemp())
+
+        def touch(name):
```
If you use [`path.py`](https://pythonhosted.org/path.py/api.html), this and other useful operations are built-in as methods. It would simplify this test a lot.

bdero replied:
Neat - I swapped this out in favor of path.touch.

singingwolfboy commented on openedx/core/djangoapps/public_api/courses/tests/test_course_import_export.py:
```diff
@@ -0,0 +1,416 @@
+            )
+        )
+
+        self.assertEquals(json.loads(resp_status.content)["ImportStatus"], -2)
```
Might want to break this up, so that if this test fails, it's easier to see where it fails.
```python
obj = json.loads(resp_status.content)
self.assertIn("ImportStatus", obj)
self.assertEquals(obj["ImportStatus"], -2)
```

bdero replied:
Good idea - adding this change in.

singingwolfboy commented on openedx/core/djangoapps/public_api/courses/tests/test_course_import_export.py:
```diff
@@ -0,0 +1,416 @@
+            args = {"name": self.good_tar, "course-data": [gtar]}
+            resp = self.client.post(self.url, args)
+
+        self.assertEquals(resp.status_code, 200)
```
Nitpick: you may want to use the [`assert_request_status_code`](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/courseware/tests/helpers.py#L47-L58) method that exists for exactly this sort of purpose.
```python
with open(self.good_tar) as gtar:
    kwargs = {"name": self.good_tar, "course-data": [gtar]}
    self.assert_request_status_code(200, self.url, "POST", **kwargs)
```

singingwolfboy replied:
Oh wait, that method only exists if you inherit from `LoginEnrollmentTestCase`. Never mind. (Still useful to know about, though.)

singingwolfboy commented on openedx/core/djangoapps/public_api/courses/tests/test_course_import_export.py:
```diff
@@ -0,0 +1,416 @@
+
+        return outside_tar
+
+    def _outside_tar2(self):
```
Whoa, this name is *really* confusing. Can you use a clearer name? Or better yet, unify `_outside_tar` and `_outside_tar2` into one function, that takes an optional parameter to configure how it works? Looks like the two functions are almost identical already.

bdero replied:
Agreed - I combined the functions and renamed the unsafe tar creation functions.

singingwolfboy commented on openedx/core/djangoapps/public_api/courses/tests/test_course_import_export.py:
```diff
@@ -0,0 +1,416 @@
+        JSON is unsupported.
+        """
+        resp = self.client.get(self.url, HTTP_ACCEPT='application/json')
+        self.assertEquals(resp.status_code, 200)
```
If we're testing that JSON is unsupported, why do we expect a 200 status code?

bdero replied:
Ah, this is a mismatch from when I moved the code over from the original API, which didn't support json. Will fix

singingwolfboy commented on openedx/core/djangoapps/public_api/courses/tests/test_course_import_export.py:
```diff
@@ -0,0 +1,416 @@
+        resp = self.client.get(self.url, HTTP_ACCEPT='application/x-tgz')
+        self.assertEquals(resp.status_code, 200)
+        # pylint: disable=maybe-no-member
+        self.assertIsNone(resp.get('Content-Disposition'))
```
Would it be better to say
```python
self.assertNotIn("Content-Disposition", resp)
```

bdero replied:
Yes, adding this change in

singingwolfboy commented on openedx/core/djangoapps/public_api/courses/tests/utils.py:
```diff
@@ -0,0 +1,101 @@
+    """
+    client.credentials(
+        HTTP_AUTHORIZATION="Bearer {}".format(access_token.token)
+    )
```
Instead of having this function rely on side effects, can you at least have it return the `client` object, and have functions that call it use the new `client` object that this returns?

bdero replied:
I can see how that makes sense, since it seems a little more predictable. Changed it to pass the the APIClient through.

singingwolfboy commented on openedx/core/djangoapps/public_api/courses/tests/test_course_import_export.py:
```diff
@@ -0,0 +1,416 @@
+
+@override_settings(CONTENTSTORE=TEST_DATA_CONTENTSTORE)
+class ExportTestCase(CourseTestCase):
+    """q
```
Why the random q?

bdero replied:
Whoops, well, at least it *q*'d a response? :wink:

bdero commented:
@singingwolfboy Thanks for the review. All comments addressed.

nedbat commented:
@bdero do you have a sense yet of how the Course API Unification will affect this pull request?

bdero commented:
@nedbat As the proposal is very abstract at this point, it's still difficult to tell.

bdero commented:
@ormsbee As per yesterday's architecture discussion, could you let me know where you think this API should come into the picture? I.e. where/how do you think this might "dock in" in the context of the layered API model as some of the details are being fleshed out?

ormsbee commented:
@bdero: Since import/export is mostly a conversation with the modulestore/contentstore I think in concrete terms, it would just mean:

1. Renaming the new app that this is in to indicate a narrower focus (e.g. "import_export"). Maybe in openedx/common/djangoapps/content/import_export.
2. Changing the URLs to be `/api/import_export/v1/courses/{course_key}` so that the versioning can happen at the app level, rather than for all course information.

Thank you.

kawaguchi-ks commented:
Can one of the admins verify this patch?

sarina commented:
@mhoeber this PR has documentation impact, however, it is unlikely we will have time to review it in the short term because of its size and high priority asks from many of our partners.

mhoeber commented:
Created https://openedx.atlassian.net/browse/DOC-1773

singingwolfboy commented on cms/djangoapps/contentstore/views/import_export.py:
```diff
@@ -332,92 +82,60 @@ def import_status_handler(request, course_key_string, filename=None):
+        try:
+            parent = CourseKey.from_string(parent)
+        except InvalidKeyError:
+            parent = None
```
I would expect this to trigger some kind of error response here, rather than setting `parent = None` and moving on. If someone hits this endpoint and provides a `parent` parameter, but there is some error in transforming that into a `CourseKey`, then this code will silently drop that parameter and move on -- a very difficult situation to debug! Can you make it display an error in the HTML page if that is the case? (If this was a JSON API, I'd suggest that you immediately return a 400 status code response, but you've changed this to only output HTML.)

sarina replied:
@bdero this comment is still unaddressed

bdero replied:
Thanks, I've improved this. Will post details in the review thread.

singingwolfboy commented on cms/djangoapps/contentstore/views/import_export.py:
```diff
@@ -332,92 +82,60 @@ def import_status_handler(request, course_key_string, filename=None):
+        }
+    ) + '?accept=application/x-tgz'
+
+    export_url += "&{0}".format(
```
Nitpick: rather than doing this with two different string concatenations, it might be clearer and more extensible to use a map. Something like this:
```python
queryargs = {
    "accept": "application/x-tgz",
    "redirect": reverse_course_url(
        "export_handler", unicode(course_key)
    ),
}
querystring = "&".join(
    "{key}={value}".format(key=urlencode(key), value=urlencode(value))
    for key, value in queryargs.items()
)
export_url += "?" + querystring
```

singingwolfboy commented on cms/djangoapps/contentstore/views/import_export.py:
```diff
@@ -332,92 +82,60 @@ def import_status_handler(request, course_key_string, filename=None):
+            "failed_module": failed_module,
+            "edit_unit_url":
+                reverse_usage_url("container_handler", parent.location)
+                if parent else "",
```
This works, but it's a bit hard to read. I generally try to only use the `X if foo else Y` construct when its short enough to all go on one line. Why don't you just use a regular `if` statement to set a variable depending on the value of `parent`, and pass that variable to the context here?

sarina replied:
I agree with this. @bdero  can you please make sure all existing comments are addressed?

sarina replied:
@bdero this comment is still unaddressed

bdero replied:
Done.

singingwolfboy commented on cms/djangoapps/contentstore/views/import_export.py:
```diff
@@ -332,92 +82,60 @@ def import_status_handler(request, course_key_string, filename=None):
     else:
-        # Only HTML or x-tgz request formats are supported (no JSON).
-        return HttpResponse(status=406)
+        return render_to_response('export.html', {
```
Since you're rendering the same template in either case, regardless of whether there's an error or not, it might be cleaner to just update the context, like this:
```python
context = {
    "context_course": course_module,
    "export_url": export_url,
}
if error:
    context["in_err"] = error
    context["unit"] = unit
    # etc... or you could use context.update({"dict": "of", "additional": "values"})
return render_to_response('export.html', context)
```

singingwolfboy commented on cms/envs/aws.py:
```diff
@@ -303,6 +303,19 @@
 X_FRAME_OPTIONS = ENV_TOKENS.get('X_FRAME_OPTIONS', X_FRAME_OPTIONS)
 
+##### OAUTH2 Provider ##############
+if FEATURES.get('ENABLE_OAUTH2_PROVIDER'):
```
Nitpick: is there a reason these options are all defined inside this block? Having them set shouldn't make any difference if the feature flag isn't turned on, but if something later on in the configuration-parsing process decides to flip this flag on, then these config values would already be defined, and things would just work. By contrast, the way the code is right now, you would have to re-execute all this logic to turn on the feature, in addition to flipping the feature flag on.

Realistically, I'm not sure if we ever would get into this situation, which is why this is just a nitpick. But I wanted to mention it, just in case.

sarina replied:
I agree with this. Thoughts?

sarina replied:
@bdero this comment is still unaddressed

bdero replied:
This is similar to how it looks in the original `lms/envs/aws.py` config. This can probably be removed in both places without introducing issues, but since it's been like this in the lms config, I'd rather not risk introducing issues for other apps that might be checking against these flags albeit incorrectly.

singingwolfboy commented on lms/envs/aws.py:
```diff
@@ -131,6 +131,10 @@
 
+# GITHUB_REPO_ROOT is the base directory
+# for course data
+GITHUB_REPO_ROOT = ENV_TOKENS.get('GITHUB_REPO_ROOT', GITHUB_REPO_ROOT)
```
There's nothing specifically dependent on Github, is there? (As opposed to Bitbucket, Gitlab, etc) This name makes me sad.

Looks like we're already using it in the codebase, though, so I guess we're stuck with it. :cry:

singingwolfboy commented on openedx/core/djangoapps/public_api/courses/tests/test_course_import_export.py:
```diff
@@ -0,0 +1,407 @@
+        self.unsafe_common_dir = path(tempfile.mkdtemp(dir=self.content_dir))
+
+    def tearDown(self):
+        shutil.rmtree(self.content_dir)
```
This would be more reliable as an [`addCleanup()`](https://docs.python.org/2/library/unittest.html#unittest.TestCase.addCleanup) step, so that it runs even if there's a failure in the `setUp` step.

sarina replied:
TIP 1: if you write a setUp() or tearDown() method, always use super() to call the parent implementation.  Pylint will flag occasions where you don't.

TIP 2: tearDown() methods can be replaced with addCleanup().  This is a testcase method you call from setUp(), to register a callback to clean things up.  This is nicer because you don't need a monolithic tearDown, they automatically don't interfere with each other, and you can even clean up partially set up tests.

To use addCleanup, change this:
```python
def setUp(self):
    self.thing = make_a_thing()

def tearDown(self):
    self.thing.destroy("now")
```
to this:
```python
def setUp(self):
    self.thing = make_a_thing()
    self.addCleanup(self.thing.destroy, "now")
```
Note that the arguments to addCleanup are a callable and the arguments to call it with.  Don't call your destroy method!

nedbat replied:
In this case, even better, use `openedx.core.lib.tempdir.mkdtemp_clean` instead of `tempfile.mkdtemp`, then you don't need a teardown or cleanup at all.

singingwolfboy commented on openedx/core/djangoapps/public_api/courses/views.py:
```diff
@@ -0,0 +1,517 @@
+
+            return self._export_error_response(
+                {
+                    "context_course": str(course_module.location),
```
Maybe a dumb question, but why are you using `str()` instead of `unicode()` here?

bdero commented:
There are complex merge conflicts cause by these: 8ad4d081fbdc024ed08cd1477380b395d78bb051 8c920b469c28c1a8962676038b8a9f755f709ca6
I'll need to rip import/export of content libraries out of the CMS and reconstruct these changes to get it all working again.

sarina commented:
I think you'll have to do a rebase to get Jenkins working on your branch

bdero commented:
@nedbat @singingwolfboy @sarina I don't think continuing minor refactors of the tests and primary logic in the import/export is in the scope of this PR as most of this code was just moved over from the contentstore app in the CMS. This PR is hard to maintain as-is, and while these are good suggestions, I think additional changes will make maintaining this needlessly harder.

I'm currently trying to reconstruct the changes from https://github.com/edx/edx-platform/commit/8ad4d081fbdc024ed08cd1477380b395d78bb051 in order to get content libraries working. Of course, with the bugs popping up around import/export recently, I suspect I may be playing catchup quite a bit.

bdero commented:
@mhoeber If I was to help draft the API documentation, what API documentation would be a good model for the format you prefer?

bdero commented:
There are a few things left to be done with this:

- [ ] Get all tests passing with content libraries
- [ ] The `import_status` state is persisted via session - it needs to be dealt with some other way (like memcached), since session support can't be assumed for clients using OAuth
- [ ] Change URLs to match the recommendation by @ormsbee
- [ ] Remove course list endpoint for now since there's currently no efficient way to do this in the platform, which needs to be addressed in a larger scope

mhoeber commented:
@bdero   we just merged an updated platform API doc in which we are consolidating all the API modules.   See:  https://github.com/edx/edx-platform/tree/master/docs/en_us/platform_api/source

The mobile subfolder is a good example of putting together the submodules.  

We are moving towards standardizing the docstrings for the API view classes.  Good examples in https://github.com/edx/edx-platform/blob/master/lms/djangoapps/mobile_api/course_info/views.py

Let me know if I can help.

sarina commented:
test this please

ormsbee commented:
@bdero: Hey, just wanted to check in and see how this was going. :-) Is there anything you wanted to talk about tomorrow?

bdero commented:
Thanks @ormsbee . Things have been slow with this, but it's just about done. There's just an issue exporting libraries. Nothing in particular that needs a discussion, I'd say.

Would it be possible for me to trigger another jenkins build? Not sure if there's access control for this, but I'll try saying what I see everyone saying: "test this please"

bdero commented:
test this please

sarina commented:
add to whitelist

sarina commented:
@bdero I added you to the whitelist which means future builds should properly trigger when you invoke that magic phrase. For your reference, rebasing your commits is another easy way to trigger a fresh build.

bdero commented:
Neat, thanks @sarina !

bdero commented:
test this please

sarina commented:
@bdero I had already started one for you, but you triggered yet another which I think caused the whole thing to fail. I'm going to trigger a new one, please let it run fully before triggering another. You should try to rebase to trigger new builds rather than running the command, when possible (you have no conflicts right now, so next time you want a build try rebasing first). Also, to see our flaky tests, visit http://bit.ly/1w91U41

sarina commented:
test this please

sarina commented:
Triggered build available at https://build.testeng.edx.org/job/edx-platform-all-tests-pr-flow/4546/. Again, please let this build run fully.

bdero commented:
I think this is ready for review again. I tested everything over and all seems to be working well. I'm unable to reproduce the remainder of the Jenkins test failures on my devstack (they're all passing).

sarina commented:
@bdero could you please rebase the PR? There are no conflicts. Being up to date, though, is helpful in diagnosing test failures.

sarina commented on cms/djangoapps/contentstore/tests/tests.py:
```diff
@@ -341,11 +341,11 @@ def test_course_key_decorator(self, course_key, status_code):
         Tests for the ensure_valid_course_key decorator.
         """
-        url = '/import/{course_key}'.format(course_key=course_key)
+        url = '/export/{course_key}'.format(course_key=course_key)
```
This seems like a very large change. Can you explain why it was made?

bdero replied:
Good question - I'm not sure why this was changed, so I changed it back. Though, whether it's import or export in this particular case shouldn't matter; they should both 404 if the course doesn't exist.

I also just realized that the reason this test was failing on Jenkins was due to a URL being wrong (silly me). I was accidentally not including this file while testing locally.

sarina commented on openedx/core/djangoapps/import_export/courses/tests/utils.py:
```diff
@@ -0,0 +1,103 @@
+# pylint: disable=E1101
```
Please don't disable warnings like this. Pylint warnings are each identified by a alphanumeric code (C0112) and a symbolic name (empty-docstring). Please use the symbolic names in new code or when updating existing code.

Please scope your suppressions as narrowly as possible. Prefer line-level suppressions the majority of the time.

sarina commented on openedx/core/djangoapps/import_export/courses/tests/utils.py:
```diff
@@ -0,0 +1,103 @@
+        Create a non-staff user, log them in (if authenticate=True), and return
+        the client, user to use for testing.
+        """
+        # pylint: disable=unused-variable
```
You can prefix variable names with an underscore to indicate you know they are unused. Please do that rather than disable this warning.

If you absolutely _must_ use disables, please do them inline so that the warning is only suppressed for that one line. With this fashion, you are disabling for the remainder of the function.

sarina commented on openedx/core/djangoapps/import_export/courses/tests/utils.py:
```diff
@@ -0,0 +1,103 @@
+        api_client = APIClient()
+
+        # Register an OAuth2 Client
+        # pylint: disable=unused-variable
```
You can prefix variable names with an underscore to indicate you know they are unused. Please do that rather than disable this warning.

If you absolutely _must_ use disables, please do them inline so that the warning is only suppressed for that one line. With this fashion, you are disabling for the remainder of the function.

sarina commented on openedx/core/djangoapps/import_export/courses/views.py:
```diff
@@ -0,0 +1,529 @@
+    authentication_classes = (OAuth2Authentication, SessionAuthentication)
+    permission_classes = (IsAuthenticated, IsCourseStaff)
+
+    # pylint: disable=unused-argument
```
You can prefix variable names with an underscore to indicate you know they are unused. Please do that rather than disable this warning.

If you absolutely _must_ use disables, please do them inline so that the warning is only suppressed for that one line. With this fashion, you are disabling for the remainder of the function.

sarina commented on openedx/core/djangoapps/import_export/courses/views.py:
```diff
@@ -0,0 +1,529 @@
+        else:
+            return JsonResponse(params)
+
+    # pylint: disable=unused-argument
```
You can prefix variable names with an underscore to indicate you know they are unused. Please do that rather than disable this warning.

If you absolutely _must_ use disables, please do them inline so that the warning is only suppressed for that one line. With this fashion, you are disabling for the remainder of the function.

sarina commented on openedx/core/djangoapps/import_export/courses/views.py:
```diff
@@ -0,0 +1,529 @@
+                )
+
+            logging.debug(
+                u'tar file being generated at {0}'.format(export_file.name)
```
Please use lazy logging: http://stackoverflow.com/a/4149231/2669626

sarina commented on openedx/core/djangoapps/import_export/courses/views.py:
```diff
@@ -0,0 +1,529 @@
+                    parent = modulestore().get_item(parent_loc)
+                    if parent.location.category == 'vertical':
+                        unit = parent
+            except:  # pylint: disable=bare-except
```
I prefer doing
`except Exception:  # pylint: disable=broad-except`

It's a nit, but you're explicitly disabling the more correct thing.

sarina commented on openedx/core/djangoapps/import_export/courses/views.py:
```diff
@@ -0,0 +1,529 @@
+        response['Content-Length'] = os.path.getsize(export_file.name)
+        return response
+
+    # pylint: disable=unused-argument
```
You can prefix variable names with an underscore to indicate you know they are unused. Please do that rather than disable this warning.

sarina commented on openedx/core/djangoapps/import_export/courses/views.py:
```diff
@@ -0,0 +1,529 @@
+                return JsonResponse(
+                    {
+                        'error_message': _(
+                            'We only support uploading a .tar.gz file.'
```
This is explicitly shown to the user, yes? I want to make sure you're only i18n'ing messages that are shown to the user.

bdero replied:
Yes, it's shown to users.

Could you let me know of this error formatting is okay?: https://github.com/edx/edx-platform/pull/6190/files#diff-49938df912930fb450d16ae2b949a463R462

sarina commented on openedx/core/djangoapps/import_export/courses/views.py:
```diff
@@ -0,0 +1,529 @@
+            )
+
+            new_location = courselike_items[0].location
+            logging.debug('new course at {0}'.format(new_location))
```
lazy logging

sarina commented on openedx/core/djangoapps/import_export/courses/views.py:
```diff
@@ -0,0 +1,529 @@
+            logging.debug('new course at {0}'.format(new_location))
+
+            log.info(
+                "Course import {0}: Course import successful".format(courselike_key)
```
lazy logging

sarina commented on openedx/core/djangoapps/import_export/courses/views.py:
```diff
@@ -0,0 +1,529 @@
+            self._save_request_status(request, courselike_string, 4)
+
+        # Send errors to client with stage at which error occurred.
+        except Exception as exception:   # pylint: disable=W0703
```
pylint warnings are each identified by a alphanumeric code (C0112) and a symbolic name (empty-docstring). Prefer the symbolic names in new code or when updating existing code.

sarina commented on openedx/core/djangoapps/import_export/courses/views.py:
```diff
@@ -0,0 +1,529 @@
+            if course_dir.isdir():
+                shutil.rmtree(course_dir)
+                log.info(
+                    "Course import {0}: Temp data cleared".format(courselike_key)
```
lazy logging

sarina commented:
```
15:22:56 Quality Report: pep8
15:22:56 -------------
15:22:56 ./cms/djangoapps/contentstore/views/import_export.py:122:5: E303 too many blank lines (2)
15:22:56 ./openedx/core/djangoapps/import_export/courses/views.py:467:33: E131 continuation line unaligned for hanging indent
15:22:56 ./openedx/core/djangoapps/import_export/courses/tests/test_course_import_export.py:470:43: W292 no newline at end of file
15:22:56 -------------
15:22:56 Violations: 3
15:22:56 FAILURE: pep8 count should be 0
```

sarina commented:
@bdero please address all comments and rebase this branch. Then I will send it to others for more review.

ormsbee commented:
@sarina: I'll pick up the next review round, but I might not be able to get to it until tomorrow night.

sarina commented:
@ormsbee I'd like @bdero to do the things I've asked for before passing off to another reviewer. There are a bunch of other things for you to review tomorrow night.

bdero commented:
All comments addressed.

sarina commented on openedx/core/djangoapps/import_export/courses/views.py:
```diff
@@ -0,0 +1,526 @@
+                    {
+
+                        'error_message': _(
+                            'Could not find the {0} file in the'
```
This works in Python, but not javascript (in JS, the "gettext" must be on the same line as the string). However, I think splitting the string here doesn't make sense - it's not a long string, and you're going to end up with "thepackage" as one word. Just put the whole string together on one line.

Also for strings that are translated, please don't use positional variables. Named variables give a richer context. See http://edx.readthedocs.org/projects/edx-developer-guide/en/latest/internationalization/i18n.html#style-guidelines for guidelines.

bdero replied:
Thanks, fixed.

sarina commented:
@bdero I'm not sure why your build is failing, but i think the CI is having some issues because I can't access the build graphs. It may be quality; you can try running `paver run_quality` locally to see if any errors report. Otherwise, we can get test eng to look tomorrow.

I'm out of the office until Monday. If @ormsbee can't help get the issues uncovered, I'll help out when I'm back.

sarina commented:
test this please

bdero commented:
According to the build output, quality is passing: https://build.testeng.edx.org/job/edx-platform-all-tests-pr-flow/4924/console

It looks like it's the same tests failing as before (https://build.testeng.edx.org/job/edx-platform-test-subset/46425/ , https://build.testeng.edx.org/job/edx-platform-test-subset/46421/).

sarina commented:
Good it ran this time. The last builds crapped out.

Those failures look suspiciously related to your branch. Perhaps Dave can
help out with them.
On Apr 29, 2015 5:56 PM, "Brandon DeRosier" <notifications@github.com>
wrote:

> According to the build output, quality is passing:
> https://build.testeng.edx.org/job/edx-platform-all-tests-pr-flow/4924/console
>
> It looks like it's the same tests failing as before (
> https://build.testeng.edx.org/job/edx-platform-test-subset/46425/ ,
> https://build.testeng.edx.org/job/edx-platform-test-subset/46421/).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/6190#issuecomment-97600004>.
>

bdero commented:
@sarina Those are the tests that are passing locally for me, but strangely failing on Jenkins.

As per @singingwolfboy's comments, I changed the logic for handling bad parent keys and added generic error message dialogs to the export template.
Here's what it looks like when an invalid parent is passed to the view (notice the URL):
![invalid parent screenshot](http://i.imgur.com/lycIw6B.png)
Let me know if this looks good.

All comments addressed.

sarina commented:
@bdero the interface looks good to me. :+1: (but @ormsbee will be the final reviewer and merger of this work).

The tests must pass on Jenkins before we can merge. If they are consistently failing on Jenkins, there must be some sort of issue with them.

sarina commented:
@bdero could you please rebase your branch, actually? Jenkins had some weird behaviors last week that's been fixed (also we increased the number of bok-choy shards). So please rebase ASAP, and ping me to let me know. Then I'll look into the test failures.

sarina commented:
For [this test failure](https://build.testeng.edx.org/job/edx-platform-all-tests-pr-flow/4983/testReport/junit/(root)/Course%20export%20_%20User%20is%20directed%20to%20unit%20with%20bad%20XML%20when%20export%20fails/And_I_can_click_to_go_to_the_unit_with_the_error/), I suspect the css selector isn't given enough time to populate. For example in the test there is a `world.wait_for_visible(".button.action-primary")` - is that the right thing to wait for?

bdero commented:
@sarina I'm actually able to consistently reproduce the library export failure now. It only fails during LMS testing, but not CMS. As for the acceptance test failure, I'll do some more experimenting with it keeping your suggestions in mind as well as rebase.

Could I trouble you to take a look at this @antoviaque? I'm having difficulty determining why I'm able to export courses in the context of the LMS, but can't seem to do the same for content libraries without failure, and I feel you probably have a better handle on this than me. The exception runs [deep into the modulestore](https://build.testeng.edx.org/job/edx-platform-all-tests-pr-flow/4983/testReport/openedx.core.djangoapps.import_export.courses.tests.test_course_import_export/ImportTestCase/test_library_import/).

sarina commented:
@bdero thanks for the update. Hopefully Xavier can help. Ping if you want some eyes on the acceptance test: I can ask one of our test engineers to look at it, although that may take a bit as they're quite busy.

antoviaque commented:
@bdero @sarina Thanks for letting me know about this, happy to help - this looks like a useful PR btw, thanks for doing this! 

Ideally I'd like to properly schedule time to look at this - would it be ok in terms of timing if I include it in our sprint next week? I would set aside time to get you support/review as necessary throughout the week to help you resolve library-related issues with this PR. @Kelketek would likely be the right person to look at this since he wrote the library import/export code, but he's currently fairly busy with another feature.

sarina commented:
Thats fine by me - @ormsbee still needs to review this PR as well.

On Wed, May 6, 2015 at 2:22 AM, Xavier Antoviaque <notifications@github.com>
wrote:

> @bdero <https://github.com/bdero> @sarina <https://github.com/sarina>
> Thanks for letting me know about this, happy to help - this looks like a
> useful PR btw, thanks for doing this!
>
> Ideally I'd like to properly schedule time to look at this - would it be
> ok in terms of timing if I include it in our sprint next week? I would set
> aside time to get you support/review as necessary throughout the week to
> help you resolve library-related issues with this PR. @Kelketek
> <https://github.com/Kelketek> would likely be the right person to look at
> this since he wrote the library import/export code, but he's currently
> fairly busy with another feature.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/6190#issuecomment-99341185>.
>

antoviaque commented:
Sounds good then - I'm adding this to the top of our backlog for next week.

bdero commented:
Thanks @antoviaque!

Kelketek commented:
@bdero The error message says that it couldn't find a library item on the draft branch: `library-v1:org7489+lib7489+branch@draft-branch`. Libraries do not have drafts. See https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/modulestore/split_mongo/split_draft.py#L229 on the master branch.

nedbat commented on openedx/core/djangoapps/import_export/courses/tests/test_course_import_export.py:
```diff
@@ -0,0 +1,484 @@
+                target_id=lib_key
+            )
+        finally:
+            shutil.rmtree(extract_dir)
```
You can use .addCleanup in test methods as well as in setUp.

singingwolfboy replied:
You could also use [`openedx.core.lib.tempdir.mkdtemp_clean`](https://github.com/edx/edx-platform/blob/master/openedx/core/lib/tempdir.py#L9-L13) instead.

nedbat replied:
Also, btw, there's a `mkdtemp_clean` function that does the clean up automatically.

bdero replied:
Alright, I changed the all of the temp directories to use mkdtemp_clean.

nedbat commented on openedx/core/djangoapps/import_export/courses/tests/test_course_import_export.py:
```diff
@@ -0,0 +1,484 @@
+            self.assertEqual(video_xml.tag, 'video')
+            self.assertEqual(video_xml.get('youtube_id_1_0'), youtube_id)
+        finally:
+            shutil.rmtree(root_dir / name)
```
this should remove root_dir, not root_dir/name, no?

nedbat commented on openedx/core/djangoapps/import_export/courses/views.py:
```diff
@@ -0,0 +1,525 @@
+                redirect_url=redirect_url
+            )
+        finally:
+            shutil.rmtree(root_dir / name)
```
Doesn't this also have to remove root_dir?

bdero replied:
Hmm, I think it should be being deleted. Just changed it to delete `root_dir`.

nedbat commented:
I've been making comments, but I'm really just nibbling at the edges.  @bdero: are there particular concerns you would like reviewers to look at?

sarina commented:
@bdero did you get the assistance you required on this pull request? Ned also made some inline comments that aren't yet addressed. Finally it will need to be rebased.

I'm trying to get Dave's attention to reviewing this pull request soon, but I want to make sure that you're also still interested and able to work on this.

bdero commented:
@sarina I'm still interested - my priorities were shifted toward CCX, but I think that's coming to a close, so I'll hopefully be able to address lingering comments shortly. Rebasing right now.

sarina commented:
@bdero good to know! I've pinged @ormsbee on the JIRA ticket and I'll harass him in person next time I see him :)

ormsbee commented:
I'll have a full pass review done before I go to sleep tonight. Thank you for your patience.

ormsbee commented on cms/djangoapps/contentstore/views/import_export.py:
```diff
@@ -2,485 +2,166 @@
-
-    elif 'text/html' in requested_format:
-        return render_to_response('export.html', context)
+            parent = CourseKey.from_string(parent)
```
In general, please avoid using the same variable to hold different types of objects. It can make it confusing to debug later on, particularly if a variable changes its type in a conditional and may be different things further on in the code. Please just use a new var name, like `parent_key`.

bdero replied:
Good idea, fixing this. Thanks.

ormsbee commented on cms/djangoapps/contentstore/views/import_export.py:
```diff
@@ -2,485 +2,166 @@
+                    "valid course or library key."
+                ).format(supplied_key=parent),
+                "library": library
+            })
```
Should this error be logged?

bdero replied:
Yes, I think this message should probably be logged, as this should never happen while trying to export courses by clicking through the UI. I'll add that in.

bdero replied:
Added in now.

ormsbee commented on cms/envs/aws.py:
```diff
@@ -43,6 +43,7 @@
 
 EMAIL_BACKEND = 'django_ses.SESBackend'
 SESSION_ENGINE = 'django.contrib.sessions.backends.cache'
+DEFAULT_FILE_STORAGE = 'storages.backends.s3boto.S3BotoStorage'
```
This might no longer be appropriate in light of #8406 -- please check?

bdero replied:
Ah yes, thanks for catching that, this was a mistake during my most recent rebase.

ormsbee commented on cms/templates/import.html:
```diff
@@ -115,7 +115,7 @@ <h3 class="title">${_("Uploading")}</h3>
             <li class="item-progresspoint item-progresspoint-unpack is-started">
               <span class="deco status-visual">
-                <i class="icon fa fa-cog"></i>
+                <i class="icon fa fa-cog fa-spin"></i>
```
Why this change?

bdero replied:
Hmm, I'm actually not sure - I can't remember if I added that in, but most likely it was an accident while performing one of the hard rebases. The import form in master seems to look and behave fine, so I'll get rid of it.

ormsbee commented on lms/envs/common.py:
```diff
@@ -589,6 +591,12 @@
+COURSELIKE_KEY_PATTERN = r'(?P<course_key_string>({}|{}))'.format(
+    r'[^/:+]+/[^/:+]+/[^/:+]+',
+    r'[^/:]+:[^/+]+\+[^/+]+(\+[^/]+)?',
+)
```
Wait, did this change from the old value?

bdero replied:
Yes, taking some commits from before I rebased & squashed, the original move is here: https://github.com/mitocw/edx-platform/commit/b6413e8def359123d851da0e6e2146133159d1c7

And the change to the regex itself is here: https://github.com/mitocw/edx-platform/commit/8c80b47996b70c2d8c9be365c873045a1065c1ff

Note the explanation:
> This fixes some instances of incorrect matching of org/course/run
> keys. For example:
> "library-v1:MITx+TL/import_status/library.3HtOcx.tar.gz" was being
> matched as a course key, even though the valid key is
> "library-v1:MITx+TL" in this case.
>
> Studio doesn't let users use `+` or `:` symbols when filling out the
> org/course/run for new courses, so I just factored that into the
> org/course/run pattern.

ormsbee replied:
@cpennington, @doctoryes: Just wanted to give you a heads up that course key URL pattern is changing, in case it raises any alarm bells.

cpennington replied:
I don't like that this changes `COURSELIKE_KEY_PATTERN` without changing `COURSE_KEY_PATTERN`. It seems like the bug (matching the additional slashes after a `library-v1`) would also affect `course-v1` keys that are matched with the `COURSE_KEY_PATTERN`.

ormsbee commented on cms/envs/common.py:
```diff
@@ -200,6 +203,29 @@
+    'oauth2_provider.oidc.handlers.ProfileHandler',
+    'oauth2_provider.oidc.handlers.EmailHandler',
+    'oauth2_handler.UserInfoHandler'
+)
```
@e0d: Config changes that will impact deployment, as discussed.

ormsbee commented on lms/envs/aws.py:
```diff
@@ -133,6 +133,10 @@
 
+# GITHUB_REPO_ROOT is the base directory
+# for course data
+GITHUB_REPO_ROOT = ENV_TOKENS.get('GITHUB_REPO_ROOT', GITHUB_REPO_ROOT)
```
Why was this added (as opposed to `COURSES_ROOT`)? That's a genuine question -- I don't understand the relationship, or where this was defined before (I notice that it existed in code that was moved?).

bdero replied:
Good question, I'm actually not sure what the difference is - they're both set to `ENV_ROOT / "data"` by default. I just moved it over because that's what the original api used:

```grep
./cms/djangoapps/contentstore/views/import_export.py:104:                data_root = path(settings.GITHUB_REPO_ROOT)
./cms/djangoapps/contentstore/views/import_export.py:270:                        settings.GITHUB_REPO_ROOT, [dirpath],
./cms/djangoapps/contentstore/views/tests/test_import_export.py:303:                settings.GITHUB_REPO_ROOT, [extract_dir / 'library'],
```

ormsbee replied:
Okay.

ormsbee commented on lms/urls.py:
```diff
@@ -91,6 +91,8 @@
     url(r'^api/val/v0/', include('edxval.urls')),
 
+    # Full Course/Library Import/Export API
+    url(r'^api/import_export/v1/', include('openedx.core.djangoapps.import_export.urls')),
```
Why is this in the LMS urls.py?

bdero replied:
So that the API is accessible from both LMS and CMS. That was one of the things I thought would be controversial, but I also don't think this is not strictly necessary. It just makes us able to export courses using this API residentially if we don't want to run studio in production. (Thoughts on this @carsongee ?) For imports/exports with edx.org it isn't necessary.

ormsbee replied:
My big concern with it is just that the modulestore as configured for the LMS doesn't currently allow you to write to content or settings scopes.

carsongee replied:
If it is controversial, then I'm happy to have it behind a feature flag, but it would be extremely helpful in automating our production processes which we share much of with edx.org, i.e. migrate a course from courses.edge to courses.edx.org, but we disable studio on our production courses site to save on resources (which edx.org could potentially do more easily with this change if they so desired).

carsongee replied:
I'm not super familiar with those scopes (and what is in them we would be missing), but we are already importing courses on the LMS with the sysadmin dashboard course import feature, which has been in for a really long time, this would just expose that as a more easily automatable endpoint.

bdero replied:
Thanks Carson, and just to append to that - this API seems to be working fine in LMS (and the relevant tests pass).

ormsbee replied:
So Cale reminded me that the write restrictions happen at the field data layer and that this is reaching straight to the modulestore. In which case I think this is fine, but please hide it behind a feature flag and have the default behavior be that it does not appear in the LMS.

ormsbee commented on openedx/core/djangoapps/import_export/courses/views.py:
```diff
@@ -0,0 +1,530 @@
+        cache.set(
+            "{}|{}".format(request.user.username, key),
+            status
+        )
```
1. Please namespace your key (e.g. `import_export.import.status:{}|{key}`) If the import crashes out in the middle, is this value cleared somehow?
2. Will this value get cleaned up if the import dies part way through (this has happened with distressing frequency as of late)?

bdero replied:
1. All set.
2. Yep, the import status is set to an error state right before logging/cleanup if any unhandled exceptions are thrown: https://github.com/edx/edx-platform/pull/6190/files#diff-49938df912930fb450d16ae2b949a463R391

bdero replied:
The above diff reference is drifting away from it's initial position due to changes I've been making to address other comments. The lines I'm talking about are:
```python
except Exception as exception: # pylint: disable=broad-except
    self._save_request_status(request, courselike_string, -1)
```

ormsbee commented on openedx/core/djangoapps/import_export/courses/views.py:
```diff
@@ -0,0 +1,530 @@
+                )
+            # set failed stage number with negative sign in case of an
+            # unsuccessful import
+            if cache.get(status_key) != 4:
```
So I realize you're only adapting existing code, but could you please assign these stage names some constants? It's low risk, and it's crazy to me that we actually labeled the steps 1, 2, 3, and 4 as opposed to giving them meaningful names.

sarina replied:
It's low risk but wouldn't it be safer, and also easier to validate the correctness, if this refactoring was done separately to this PR?

ormsbee replied:
... ok. I guess I just had a reflex twitch at seeing this, but I take your point. I'm fine with putting this off -- it doesn't make the codebase any worse than it already is, and this is a tangly part of the codebase.

bdero replied:
There is also status 0 and -1. ;) I agree that these should be a set of constants, and also that we should do this in another PR.

cpennington replied:
It appears I just replicated this comment earlier in the PR.

ormsbee commented:
Done with this pass, sorry for the delay. Aside from the items I commented on, I had a few higher level concerns:

1. Can you please describe how the manual testing has been done? Does it include old mongo, split, content libraries? I ask because this is a traditionally troublesome part of the codebase for us. It's possible that if you have a plan of what should be tested, we might be able to give some QA support (that's not definite -- I have to check).
2. It would be worthwhile to re-broadcast this loudly to edx-code. The old import mechanism was never meant as a public API, but it's quite possible that others have built tooling around it. Describing what changes are necessary to use this system would be very helpful. I imagine you've already thought that through anyway, since your own scripts are going to make the transition.

Thank you.

bdero commented:
1. I need to do a detailed round of manual testing for this again, since there have been a lot of little changes since the last time. But yes, I did test course imports with old mongo & split, as well as content libraries. For testing manually, I have an exported split and old mongo course, which I import using the API, then reexport and compare the differences.
2. Sounds good, here's a draft of what I would mention: Everything should appear and work pretty much exactly as it does now from the perspective of studio, except that there is a minor change to the import view; it displays a specific error message when a given course key is invalid (which was a change discussed in this PR). Any existing tools would need to at least change the import and export URLs. Other than that, exporting via session authentication will still work the same as it does now. However, I recommend authenticating via OAuth2 (...then goes on to explain how you'd do that...).

bdero commented:
I think I addressed all comments from @ormsbee - I will get the comments from Ned and DB in a few. Looking to see what the test results will be for this next one.

bdero commented:
Ok, addressed comments by @nedbat - looking into test failures on Jenkins

bdero commented:
# Test failures

### [lms.djangoapps.verify_student.tests.test_views.TestEmailMessageWithCustomICRVBlock.test_denied_email_message_with_valid_due_date_and_attempts_allowed](https://build.testeng.edx.org/job/edx-platform-all-tests-pr-flow/10070/testReport/lms.djangoapps.verify_student.tests.test_views/TestEmailMessageWithCustomICRVBlock/test_denied_email_message_with_valid_due_date_and_attempts_allowed/):
I was able to reproduce this locally, but it looks both unrelated and this same test failing for everyone's builds right now, so I'll stand by for now on this test.
##### Examples of other PR failures:
https://build.testeng.edx.org/job/edx-platform-all-tests-pr-flow/10014/testReport/
https://build.testeng.edx.org/job/edx-platform-all-tests-pr-flow/10019/testReport/
https://build.testeng.edx.org/job/edx-platform-all-tests-pr-flow/10022/testReport/

### All of the `JavaScript.firefox.*` failures:
Investigating..

sarina commented:
Brandon - the ICRV failure is known. It will be fixed in master at some
point soon. The firefox failures tend to just be flaky. I recommend
rebasing your branch and getting a fresh build.

On Mon, Jun 22, 2015 at 2:57 PM, Brandon DeRosier <notifications@github.com>
wrote:

> Test failures
> lms.djangoapps.verify_student.tests.test_views.TestEmailMessageWithCustomICRVBlock.test_denied_email_message_with_valid_due_date_and_attempts_allowed
> <https://build.testeng.edx.org/job/edx-platform-all-tests-pr-flow/10070/testReport/lms.djangoapps.verify_student.tests.test_views/TestEmailMessageWithCustomICRVBlock/test_denied_email_message_with_valid_due_date_and_attempts_allowed/>
> :
>
> I was able to reproduce this locally, but it looks both unrelated and this
> same test failing for everyone's builds right now, so I'll stand by for now
> on this test.
> Examples of other PR failures:
>
>
> https://build.testeng.edx.org/job/edx-platform-all-tests-pr-flow/10014/testReport/
>
> https://build.testeng.edx.org/job/edx-platform-all-tests-pr-flow/10019/testReport/
>
> https://build.testeng.edx.org/job/edx-platform-all-tests-pr-flow/10022/testReport/
> All of the JavaScript.firefox.* failures:
>
> Investigating..
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/6190#issuecomment-114221207>.
>

bdero commented:
Ok, I've been making updates to my test failures comment as a progress. I should be down to only one more test failure:

### [Course export : User is directed to unit with bad XML when export fails.And I can click to go to the unit with the error](https://build.testeng.edx.org/job/edx-platform-all-tests-pr-flow/10079/testReport/%28root%29/Course%20export%20_%20User%20is%20directed%20to%20unit%20with%20bad%20XML%20when%20export%20fails/And_I_can_click_to_go_to_the_unit_with_the_error/):
- Need to reproduce locally. Will work on this tomorrow.

bdero commented:
Ok, that should do it I think. I just pushed the fix for the last acceptance test. All comments have been addressed, and this is ready for continued review. @sarina @ormsbee

sarina commented:
@bdero please do rebase as well to pick up the ICRV test fix.

bdero commented:
@sarina Just rebased.

bdero commented:
Just fixed the 4 tiny quality issues, hopefully this will have a green light from Jenkins this time around.

bdero commented:
Build passing! :) I'm constructing a message which I will post to the code list about this.

bdero commented:
I posted a message to the code list earlier today as requested: https://groups.google.com/forum/#!topic/edx-code/6dP9SEKsmqQ

ormsbee commented:
Did you see @jzoldak's email question about whether it would impact the import management command? Just want to make sure it's on your testing radar.

ormsbee commented:
Items before merging:

1. Feature flag so that it is only enabled in Studio by default.
2. Verification that management command driven imports still work.
3. Test plan and sample data that we can give to our manual testing team.
4. Squash.
5. Profit!

Thank you.

cpennington commented on openedx/core/djangoapps/import_export/courses/views.py:
```diff
@@ -0,0 +1,528 @@
+        except InvalidKeyError:
+            raise Http404
+
+        return has_course_author_access(request.user, course_key)
```
This appears to be a deprecated function (it just redirects to `has_studio_write_access`). That in itself is somewhat misleading, since we're making this not a studio-only thing anymore. @ormsbee, @nasthagiri: This appears to be a prime candidate for any upcoming roles/permissions improvements.

ormsbee replied:
Wasn't that in the old code as well though?

FWIW, I want to be very careful about introducing anything new in this PR. Import is complicated enough -- just moving it as much as this is risky. I don't want to tag on more general refactoring on this yet.

cpennington replied:
Yeah, makes sense.

cpennington commented on openedx/core/djangoapps/import_export/courses/views.py:
```diff
@@ -0,0 +1,528 @@
+)
+
+
+class IsCourseStaff(BasePermission):
```
It would be better to name this something like `HasCourseWriteAccess` (to distinguish the role (staff) from the permission (write access)).

bdero replied:
Thanks, I've changed this.

cpennington commented on openedx/core/djangoapps/import_export/courses/views.py:
```diff
@@ -0,0 +1,528 @@
+        # properly cleaned up.
+        try:
+            # Cache the import progress
+            self._save_request_status(request, courselike_string, 0)
```
Since this is just shown back to a particular user, maybe it should just go into the user's session, rather than directly into the cache?

ormsbee replied:
In the OAuth2/API use case, they don't necessarily have a session.

cpennington commented on openedx/core/djangoapps/import_export/courses/views.py:
```diff
@@ -0,0 +1,528 @@
+            1 : Extracting file
+            2 : Validating.
+            3 : Importing to mongo
+            4 : Import successful
```
Can you put these values into an enum, to give them explicit names?

ormsbee replied:
We covered that earlier in the conversation. Just transplanting what was there in this PR.

bdero commented:
Ok, going to start checking things off the checklist:

- [ ] 1.  Feature flag so that it is only enabled in Studio by default.
- [ ] 2. Verification that management command driven imports still work.
- [ ] 3. Test plan and sample data that we can give to our manual testing team.
- [ ] 4. Squash.
- [ ] 5. Profit!

bdero commented:
@ormsbee I Added a feature flag (in a separate commit ATM to make review easier). Let me know if that suffices for that change. I think the only things left are 3 and 5!

bdero commented:
Also, here's an outline of what I think a good testing plan for this might be:

1. Login to Studio
2. Create a new course.
3. Populate it with 1 chapter, 1 sequential, 1 vertical, 1 unit, and 1 multiple-choice problem.
4. Publish the content
5. Export the content tarball using Studio's Export view
6. Reimport the content using Studio's Import view, verifying that the import status checklist live updates as the content import progresses.
7. Create a new Advanced Problem and populate it with invalid OLX.
8. Publish the content.
9. Again, navigate to Studio's export view and try to export the content, verifying that there is an error.
10. Click the first button on the error dialog and verify that it leads to a view of the unit containing the problem with broken OLX.
11. Create a rerun of the course (for testing against Split).
12. Repeat steps 4-10 using the course rerun.
13. Create a new Content Library.
14. Repeat steps 3-10 using the Content Library.
15. ...[steps for testing OAuth]...

bdero commented:
Hmm, looks like overriding the new feature flag on the base test class didn't fare out so well on jenkins. I'll just enable it in the test environment, which is probably the better way to do it anyways.

sarina commented:
@bdero how did you override it? I find both of these work well:

`with mock.patch.dict('django.conf.settings.FEATURES', {'ALLOW_UNICODE_COURSE_ID': False}):`

or as a decorator on the test class itself (either per test_ method, or on the setup method):
`@patch.dict(settings.FEATURES, {'ENABLE_EXPORT_GIT': True})`

bdero commented:
@sarina It was using override_settings as a class decorator here, which seemed to work locally for me: https://github.com/mitocw/edx-platform/commit/eefebb0d156fbd37c247407d1a3bd0002814952b#diff-f2f4619ac1fc8a05f4c3f9f438fef6f3R70

Just pushed my second attempt and it looks like I have fresh merge conflicts.. Going to rebase.

bdero commented:
Just rebased. It was just a tiny conflict in the test env.

@ormsbee How should we coordinate on creating a test plan?

bdero commented:
All is well with the build again! :)

bdero commented:
Is there anything I can do to help get this moving again? It would be really awesome to have this in Cypress if possible since we follow named releases and have a residential use case for it.

sarina commented:
@bdero yeah I'm trying to get @ormsbee to take a look at this today. We are cutting Cypress on monday so it is a bit tight here.

ormsbee commented:
@bdero: Hey, sorry for the delay. Please just squash and give a descriptive commit message (you can take it largely from your PR summary), and I'll hit the merge button. Thank you for your patience.

ormsbee commented:
FWIW, I'm fine with keeping these as two separate commits, I just want a more descriptive message for them, since it's a pretty big change.

bdero commented:
Thanks, I've updated the commit messages with details. Let me know if they're detailed enough. I appreciate that you guys are getting around to this despite your busy schedules with the upcoming release.

ormsbee commented:
@bdero: looks good, ty. As soon as the build turns green, I'll hit the button.

bdero commented:
Failed on the flaky firefox ones, trying again.

test this please

bdero commented:
test this please

bdero commented:
Aborted this time.... Trying again.

bdero commented:
test this please

bdero commented:
The JS tests failed again. Going to rebase and push.

bdero commented:
:pray:

sarina commented:
@bdero FYI please don't run another build just yet, let's capture what's failing if you continue to see failures.

In this previous build: https://build.testeng.edx.org/job/edx-platform-all-tests-pr-flow/11079/BuildGraph/ you had everything passing except common-js-unit-1 shard; in your current build that shard has already passed. So you should be cleared to merge now, if there are other failures let's PAUSE and make sure they're logged as flaky.

sarina commented:
See https://openedx.atlassian.net/wiki/display/TE/Flaky+Test+Process FYI. Flagging flaky tests as flaky helps other developers, and you.

sarina commented:
whew passing build! I'm definitely going to leave the merging honors to @ormsbee

bdero commented:
Woohoo! Thank you!

sarina commented:
@mhoeber it occurs to me that this PR will need documentation. It's been open longer than our Documentation needs checklist has existed so it slipped from my mind until now. Sorry about that. @ormsbee and @bdero can answer any questions you may have.
